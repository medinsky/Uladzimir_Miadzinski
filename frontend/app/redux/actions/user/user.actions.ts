import { Action } from '@ngrx/store';
import { User, UserCredentials } from '../../../user-list/user-service.interface';

export const LOGIN_USER = '[User] Login User';
export const LOGIN_USER_FAIL = '[User] Login User Fail';
export const LOGIN_USER_SUCCESS = '[User] Login User Success';
export const LOGOUT_USER = '[User] Logout User';
export const LOGOUT_USER_FAIL = '[User] Logout User Fail';
export const LOGOUT_USER_SUCCESS = '[User] Logout User Success';
export const LOAD_USERS = '[User] Load Users';
export const LOAD_USERS_FAIL = '[User] Load Users Fail';
export const LOAD_USERS_SUCCESS = '[User] Load Users Success';
export const CREATE_USER = '[User] Create User';
export const CREATE_USER_FAIL = '[User] Create User Fail';
export const CREATE_USER_SUCCESS = '[User] Create User Success';
export const UPDATE_USER = '[User] Update User';
export const UPDATE_USER_FAIL = '[User] Update User Fail';
export const UPDATE_USER_SUCCESS = '[User] Update User Success';
export const DELETE_USER = '[User] Delete User';
export const DELETE_USER_FAIL = '[User] Delete User Fail';
export const DELETE_USER_SUCCESS = '[User] Delete User Success';
export const ASSIGN_USER_PASSWORD = '[User] Assign User Password';
export const ASSIGN_USER_PASSWORD_FAIL = '[User] Assign User Password Fail';
export const ASSIGN_USER_PASSWORD_SUCCESS = '[User] Assign User Password Success';
export const UPDATE_CURRENT_USER = '[User] Update Current User';
export const UPDATE_CURRENT_USER_FAIL = '[User] Update Current User Fail';
export const UPDATE_CURRENT_USER_SUCCESS = '[User] Update Current User Success';
export const LOAD_CURRENT_USER = '[User] Load Current User';
export const LOAD_CURRENT_USER_FAIL = '[User] Load Current User Fail';
export const LOAD_CURRENT_USER_SUCCESS = '[User] Load Current User Success';
export const RESET_DATA_STATE = '[User] Reset Data State';

export class LoadUsers implements Action {
  readonly type = LOAD_USERS;
}

export class LoadUsersFail implements Action {
  readonly type = LOAD_USERS_FAIL;

  constructor(public payload: string) {
  }
}

export class LoadUsersSuccess implements Action {
  readonly type = LOAD_USERS_SUCCESS;

  constructor(public payload: User[]) {
  }
}

export class LoginUser implements Action {
  readonly type = LOGIN_USER;

  constructor(public payload: UserCredentials) {
  }
}

export class LoginUserFail implements Action {
  readonly type = LOGIN_USER_FAIL;

  constructor(public payload: string | number) {
  }
}

export class LoginUserSuccess implements Action {
  readonly type = LOGIN_USER_SUCCESS;

  constructor(public payload: User) {
  }
}

export class LogoutUser implements Action {
  readonly type = LOGOUT_USER;
}

export class LogoutUserFail implements Action {
  readonly type = LOGOUT_USER_FAIL;

  constructor(public payload: string | number) {
  }
}

export class LogoutUserSuccess implements Action {
  readonly type = LOGOUT_USER_SUCCESS;
}

export class CreateUser implements Action {
  readonly type = CREATE_USER;

  constructor(public payload: User) {
  }
}

export class CreateUserFail implements Action {
  readonly type = CREATE_USER_FAIL;

  constructor(public payload: string) {
  }
}

export class CreateUserSuccess implements Action {
  readonly type = CREATE_USER_SUCCESS;

  constructor(public payload: User) {
  }
}

export class UpdateUser implements Action {
  readonly type = UPDATE_USER;

  constructor(public payload: User) {
  }
}

export class UpdateUserFail implements Action {
  readonly type = UPDATE_USER_FAIL;

  constructor(public payload: string) {
  }
}


export class AssignUserPassword implements Action {
  readonly type = ASSIGN_USER_PASSWORD;

  constructor(public payload: UserCredentials) {
  }
}

export class AssignUserPasswordFail implements Action {
  readonly type = ASSIGN_USER_PASSWORD_FAIL;

  constructor(public payload: string) {
  }
}

export class AssignUserPasswordSuccess implements Action {
  readonly type = ASSIGN_USER_PASSWORD_SUCCESS;
}

export class UpdateUserSuccess implements Action {
  readonly type = UPDATE_USER_SUCCESS;

  constructor(public payload: User) {
  }
}

export class DeleteUser implements Action {
  readonly type = DELETE_USER;

  constructor(public payload: number | undefined) {
  }
}

export class DeleteUserFail implements Action {
  readonly type = DELETE_USER_FAIL;

  constructor(public payload: string) {
  }
}

export class DeleteUserSuccess implements Action {
  readonly type = DELETE_USER_SUCCESS;

  constructor(public payload: User) {
  }
}

export class UpdateCurrentUser implements Action {
  readonly type = UPDATE_CURRENT_USER;

  constructor(public payload: User) {
  }
}

export class UpdateCurrentUserFail implements Action {
  readonly type = UPDATE_CURRENT_USER_FAIL;

  constructor(public payload: string) {
  }
}

export class UpdateCurrentUserSuccess implements Action {
  readonly type = UPDATE_CURRENT_USER_SUCCESS;

  constructor(public payload: User) {
  }
}

export class LoadCurrentUser implements Action {
  readonly type = LOAD_CURRENT_USER;
}

export class LoadCurrentUserFail implements Action {
  readonly type = LOAD_CURRENT_USER_FAIL;

  constructor(public payload: string) {
  }
}

export class LoadCurrentUserSuccess implements Action {
  readonly type = LOAD_CURRENT_USER_SUCCESS;

  constructor(public payload: User) {
  }
}

export class ResetDataState implements Action {
  readonly type = RESET_DATA_STATE;
}

export type UsersActions =
  ResetDataState | LoadUsers | LoadUsersFail | LoadUsersSuccess |
  CreateUser | CreateUserFail | CreateUserSuccess |
  LoadCurrentUser | LoadCurrentUserFail | LoadCurrentUserSuccess |
  LoginUser | LoginUserFail | LoginUserSuccess |
  LogoutUser | LogoutUserFail | LogoutUserSuccess |
  UpdateUser | UpdateUserFail | UpdateUserSuccess |
  UpdateCurrentUser | UpdateCurrentUserFail | UpdateCurrentUserSuccess |
  DeleteUser | DeleteUserFail | DeleteUserSuccess |
  AssignUserPassword | AssignUserPasswordFail | AssignUserPasswordSuccess;
