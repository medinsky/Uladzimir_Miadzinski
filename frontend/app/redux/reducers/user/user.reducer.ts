import {
  ASSIGN_USER_PASSWORD, ASSIGN_USER_PASSWORD_FAIL, ASSIGN_USER_PASSWORD_SUCCESS,
  CREATE_USER, CREATE_USER_FAIL, CREATE_USER_SUCCESS, DELETE_USER, DELETE_USER_FAIL, DELETE_USER_SUCCESS,
  LOAD_USERS,
  LOAD_USERS_FAIL,
  LOAD_USERS_SUCCESS, RESET_DATA_STATE,
  UPDATE_USER, UPDATE_USER_FAIL, UPDATE_USER_SUCCESS,
  UsersActions
} from '../../actions/user/user.actions';
import { User } from '../../../user-list/user-service.interface';
import { Map } from 'immutable';

export interface UsersState {
  data: User[];
  loaded?: boolean;
  loading?: boolean;
}

export const initialState: UsersState = {
  data: [],
  loaded: false,
  loading: false,
};

export function userReducer(state: UsersState = initialState, action: UsersActions): UsersState {

  switch (action.type) {

    case RESET_DATA_STATE: {
      return Map(state)
        .set('data', [])
        .set('loading', false)
        .set('loaded', false)
        .toJS();
    }

    case LOAD_USERS: {
      return Map(state)
        .set('loading', true)
        .set('loaded', false)
        .toJS();
    }

    case LOAD_USERS_FAIL: {
      return Map(state)
        .set('data', action.payload)
        .set('loading', false)
        .set('loaded', false)
        .toJS();
    }

    case LOAD_USERS_SUCCESS: {
      return Map(state)
        .set('data', action.payload)
        .set('loading', false)
        .set('loaded', true)
        .toJS();
    }

    case CREATE_USER: {
      return Map(state)
        .set('loading', true)
        .set('loaded', false)
        .toJS();
    }

    case CREATE_USER_FAIL: {
      return Map(state)
        .set('loading', false)
        .set('loaded', false)
        .toJS();
    }

    case CREATE_USER_SUCCESS: {
      return Map(state)
        .set('loading', false)
        .set('loaded', true)
        .toJS();
    }

    case UPDATE_USER: {
      return Map(state)
        .set('loading', true)
        .set('loaded', false)
        .toJS();
    }

    case UPDATE_USER_FAIL: {
      return Map(state)
        .set('loading', false)
        .set('loaded', false)
        .toJS();
    }

    case UPDATE_USER_SUCCESS: {
      return Map(state)
        .set('loading', false)
        .set('loaded', true)
        .toJS();
    }

    case ASSIGN_USER_PASSWORD: {
      return Map(state)
        .set('loading', true)
        .set('loaded', false)
        .toJS();
    }

    case ASSIGN_USER_PASSWORD_FAIL: {
      return Map(state)
        .set('loading', false)
        .set('loaded', false)
        .toJS();
    }

    case ASSIGN_USER_PASSWORD_SUCCESS: {
      return Map(state)
        .set('loading', false)
        .set('loaded', true)
        .toJS();
    }

    case DELETE_USER: {
      return Map(state)
        .set('loading', true)
        .set('loaded', false)
        .toJS();
    }

    case DELETE_USER_FAIL: {
      return Map(state)
        .set('loading', false)
        .set('loaded', false)
        .toJS();
    }

    case DELETE_USER_SUCCESS: {
      return Map(state)
        .set('loading', false)
        .set('loaded', true)
        .toJS();
    }

    default: {
      return state;
    }
  }
}
