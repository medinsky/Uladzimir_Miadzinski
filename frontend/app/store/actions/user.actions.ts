import { Action } from '@ngrx/store';
import { User } from '../../user-list/user-service.interface';

export const LOAD_USERS = '[User] Load Users';
export const LOAD_USERS_FAIL = '[User] Load Users Fail';
export const LOAD_USERS_SUCCESS = '[User] Load Users Success';
export const POST_USER = '[User] Post User';
export const POST_USER_FAIL = '[User] Post User Fail';
export const POST_USER_SUCCESS = '[User] Post User Success';

export class LoadUsers implements Action {
  readonly type = LOAD_USERS;
}

export class LoadUsersFail implements Action {
  readonly type = LOAD_USERS_FAIL;
  constructor(public payload: string) {}
}

export class LoadUsersSuccess implements Action {
  readonly type = LOAD_USERS_SUCCESS;
  constructor(public payload: User[]) {}
}

export class PostUser implements Action {
  readonly type = POST_USER;
  constructor(public payload: User) {}
}

export class PostUserFail implements Action {
  readonly type = POST_USER_FAIL;
  constructor(public payload: string) {}
}

export class PostUserSuccess implements Action {
  readonly type = POST_USER_SUCCESS;
  constructor(public payload: User) {}
}

export type UsersActions = LoadUsers | LoadUsersFail | LoadUsersSuccess;

export type UserPostActions = PostUser | PostUserFail | PostUserSuccess;
