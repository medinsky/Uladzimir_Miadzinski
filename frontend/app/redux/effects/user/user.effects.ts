import { Injectable } from '@angular/core';
import { Actions, Effect, ofType } from '@ngrx/effects';
import { Observable, of } from 'rxjs';
import {
  UsersActions,
  LOAD_USERS,
  LoadUsersSuccess,
  LoadUsersFail,
  POST_USER,
  PostUser,
  PostUserSuccess,
  PostUserFail,
  LoadUsers,
  LOAD_CURRENT_USER,
  LoadCurrentUserSuccess,
  LoadCurrentUserFail,
  LOGIN_USER,
  LoginUser,
  LoginUserSuccess,
  LoginUserFail, LogoutUserSuccess, LogoutUserFail, LOGOUT_USER,
} from '../../actions/user/user.actions';
import { catchError, map, mapTo, mergeMap, switchMap } from 'rxjs/operators';
import { UserService } from '../../../services/user.service';
import { User, UserCredentials } from '../../../user-list/user-service.interface';
import { AuthService } from '../../../services/auth.service';
import { HttpErrorResponse } from '@angular/common/http';
import { STATUS } from '../../../login/login.component';

@Injectable()
export class UserEffects {

  constructor(
    private actions$: Actions,
    private userService: UserService,
    private authService: AuthService
  ) {
  }

  @Effect()
  loadUsers$: Observable<UsersActions> = this.actions$.pipe(
    ofType(LOAD_USERS),
    switchMap(() => {
      return this.userService.getUsers()
        .pipe(
          map((users: User[]) => {
            return new LoadUsersSuccess(users);
          }),
          catchError((error: string) => of(new LoadUsersFail(error)))
        );
    })
  );

  @Effect()
  loadCurrentUser$: Observable<UsersActions> = this.actions$.pipe(
    ofType(LOAD_CURRENT_USER),
    switchMap(() => {
      return this.userService.getCurrentUser()
        .pipe(
          map((user: User) => {
            return new LoadCurrentUserSuccess(user);
          }),
          catchError((error: string) => of(new LoadCurrentUserFail(error)))
        );
    })
  );

  @Effect()
  postUser$: Observable<UsersActions> = this.actions$.pipe(
    ofType(POST_USER),
    mergeMap((action: PostUser) => {
      return this.userService.createUser(action.payload as User)
        .pipe(
          map((user: User) => {
            return new PostUserSuccess(user);
          }),
          mapTo(new LoadUsers()),
          catchError((error: string) => of(new PostUserFail(error)))
        );
    })
  );

  @Effect()
  loginUser$: Observable<UsersActions> = this.actions$.pipe(
    ofType(LOGIN_USER),
    mergeMap((action: LoginUser) => {
      return this.authService.login(action.payload as UserCredentials)
        .pipe(
          map((user: User) => new LoginUserSuccess(user)),
          catchError((error: HttpErrorResponse) => {
            return of(new LoginUserFail(error.status === STATUS.UNAUTHORIZED
                ? STATUS.UNAUTHORIZED
                : error.message));
          })
        );
    })
  );

  @Effect()
  logoutUser$: Observable<UsersActions> = this.actions$.pipe(
    ofType(LOGOUT_USER),
    mergeMap(() => {
      return this.authService.logout()
        .pipe(
          map(() => new LogoutUserSuccess()),
          catchError((error: HttpErrorResponse) => of(new LogoutUserFail(error.message)))
        );
    })
  );

}


