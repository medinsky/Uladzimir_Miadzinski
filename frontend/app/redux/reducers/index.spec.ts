import { currentUserReducer } from './user/current-user.reducer';
import { LoginUser } from '../actions/user/user.actions';
import { UserCredentials } from '../../user-list/user-service.interface';

describe('Login current user with name and password', () => {
  test('should start loading', () => {
    const user: UserCredentials = {
      name: 'Medinsky',
      password: '1'
    };

    const result = currentUserReducer(undefined, new LoginUser(user));

    expect(result).toMatchSnapshot();
  });
});
