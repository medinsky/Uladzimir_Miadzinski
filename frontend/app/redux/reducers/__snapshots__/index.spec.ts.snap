// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login current user with name and password should start loading 1`] = `
Object {
  "data": Object {
    "name": "Medinsky",
    "password": "1",
  },
  "loaded": false,
  "loading": true,
}
`;
